analyzers/src/SonarAnalyzer.CSharp/Rules/AvoidExcessiveInheritance.cs

```diff
@@ -36,89 +38,108 @@ namespace SonarAnalyzer.Rules.CSharp
         private string filteredClasses = FilteredClassesDefaultValue;
         [RuleParameter(
-            key: "filteredClasses",
-            type: PropertyType.String,
-            description: "Comma-separated list of classes to be filtered out of the count of inheritance. Depth " +
+            "filteredClasses",
+            PropertyType.String,
+            "Comma-separated list of classes to be filtered out of the count of inheritance. Depth " +
             "counting will stop when a filtered class is reached. For example: System.Windows.Controls.UserControl, " +
             "System.Windows.*. (String)",
-            defaultValue: FilteredClassesDefaultValue)]
+            FilteredClassesDefaultValue)]
         public string FilteredClasses
         {
-            get => this.filteredClasses;
+            get => filteredClasses;
             set
             {
-                this.filteredClasses = value;
-                this.filteredClassesRegex = this.filteredClasses.Split(',')
-                    .Select(WilcardPatternToRegularExpression)
+                filteredClasses = value;
+                filteredClassesRegex = filteredClasses.Split(',')
+                    .Select(WildcardPatternToRegularExpression)
                     .ToList();
             }
         }
 
-        protected override void Initialize(ParameterLoadingAnalysisContext context)
-        {
+        protected override void Initialize(ParameterLoadingAnalysisContext context) =>
             context.RegisterSyntaxNodeActionInNonGenerated(
                 c =>
                 {
-                    var declaration = (ClassDeclarationSyntax)c.Node;
-                    var symbol = c.SemanticModel.GetDeclaredSymbol(declaration);
-                    if (symbol == null)
+                    var objectTypeInfo = new ObjectTypeInfo(c.Node, c.SemanticModel);
+                    // For records we are triggered twice and we need to differentiate between the calls by checking the containing symbol kind
+                    // See: https://github.com/dotnet/roslyn/issues/50989
+                    if (objectTypeInfo.Symbol == null || c.ContainingSymbol.Kind != SymbolKind.NamedType)
                     {
                         return;
                     }
 
-                    var thisTypeRootNamespace = GetRootNamespace(symbol);
+                    var thisTypeRootNamespace = GetRootNamespace(objectTypeInfo.Symbol);
 
-                    var baseTypesCount = symbol.GetSelfAndBaseTypes()
+                    var baseTypesCount = objectTypeInfo.Symbol.GetSelfAndBaseTypes()
                         .Skip(1) // remove the class itself
                         .TakeWhile(s => GetRootNamespace(s) == thisTypeRootNamespace)
                         .Select(nts => nts.OriginalDefinition.ToDisplayString())
-                        .TakeWhile(className => this.filteredClassesRegex.All(regex => !regex.IsMatch(className)))
```

Comment:
I will not change the property name since it's already used and this will be a breaking change for the users. This is not that bad since the Records are in the end immutable classes.